### ISSUE TITLE
http2 push

### ISSUE THREAD

**Issue by superpoincare** [szepeviktor/fix-w3tc#234]
https is catching up and there's server side push also if the server supports it. 

Although https is not required for push, browsers require it, so push requires both SSL support and server with the capability of pushing. 

Will be great if this can be achieved. 

This plugin seems to do it but not for the minified file: 

https://wordpress.org/plugins/http2-server-push/

Also one has to be careful pushing too many things but css pushing can be very good. Some description here: https://blog.yoav.ws/being_pushy/

**Comment by nigrosimone** [comment]
it's hard to implement for page cache enhanced but it's really good performance improvements.

**Comment by superpoincare** [comment]
I think it shouldn't be that difficult. 

[This](https://www.tunetheweb.com/performance/http2/http2-push/) suggests that the basic code is: 

```
<filesMatch "\.([hH][tT][mM][lL]?)">
   Header add Link "</assets/css/common.css>;rel=preload;as=style"
</filesMatch>
```

**Comment by nigrosimone** [comment]
w3tc page cache enhanced has a path based cache for each uri, eg.:

    http://www.mywebsite.com/my-page/

is cached in:

    /wp-content/cache/page-enhanced/www.mywebsite.com/my-page/_index.html

we can add in the same directory the .htaccess with the proper headers.

**Comment by superpoincare** [comment]
Yeah. 

One has to be careful. All pages have a different minified css output. There's a single file only if all pages have the same css files enqueued by Wordpress.

**Comment by nigrosimone** [comment]
it require a lot of option in the dashboard, i think this require a new section/sub-section for:

- allow/disallow push
- set which assets push: js, css, both (+ other media like image?)
- set max assets to push/max length for the headers 
- set max size for the assets to push: http2 push has a performance degradation is the file is to big

**Comment by amiga-500** [comment]
Very interesting.  I'll go educate myself on http/2.

I guess we could have an extension that needs to be enabled (from the extensions page) and then it adds a new subsection ("Http/2 Push") under the General Settings page.  I am slowly building a new extension for security headers so at least i am starting to become familiar with the extensions process.  But this has me more intrigued so i might put that aside and look into this instead 😆

**Comment by nigrosimone** [comment]
> I guess we could have an extension [..]

nope... i think this could be a core feature, also core wp developers will implement in the future release (4.8?). http2 will be (in the next future) the definitive http protocol version.

i'm on works, but i have very few time, i want share my basic idea.

**Step 1: store the Link preload headers in the htaccess  for cache enhanced** 

w3tc page cache enhanced has a path based cache for each uri, eg.:

    http://www.mywebsite.com/my-page/

is cached in:

    /wp-content/cache/page-enhanced/www.mywebsite.com/my-page/_index.html

we can add in the same directory the .htaccess with the proper headers.

in `/w3-total-cache-fixed/lib/W3/Cache/File/Generic.php` function `set($key, $var, $expire = 0, $group = '')`, check the headers and if a `Link` exists, store in the htaccess, eg.:

```php
    /**
     * Sets data
     *
     * @param string $key
     * @param string $var
     * @param int $expire
	 * @param string $group Used to differentiate between groups of cache values
     * @return boolean
     */
    function set($key, $var, $expire = 0, $group = '') {
        $key = $this->get_item_key($key);
        $sub_path = $this->_get_path($key);
        $path = $this->_cache_dir . '/' . $sub_path;

        $dir = dirname($path);

        if (!@is_dir($dir)) {
            if (!w3_mkdir_from($dir, W3TC_CACHE_DIR))
                return false;
        }

        $fp = @fopen($path, 'w');
        if (!$fp)
            return false;
        
        if ($this->_locking)
            @flock($fp, LOCK_EX);

        @fputs($fp, $var['content']);
        @fclose($fp);

        if ($this->_locking)
            @flock($fp, LOCK_UN);

        // some hostings create files with restrictive permissions
        // not allowing apache to read it later
        @chmod($path, 0644); 

        $old_entry_path = $path . '.old';
        @unlink($old_entry_path);

        if (w3_is_apache() && isset($var['headers']) ){
        	
        	$htaccess = '';
        	
        	if( isset($var['headers']['Content-Type']) && substr($var['headers']['Content-Type'], 0, 8) == 'text/xml' ) {
        		$htaccess .= "<IfModule mod_mime.c>\n" .
			                 "    RemoveType .html_gzip\n" .
			                 "    AddType text/xml .html_gzip\n" .
			                 "    RemoveType .html\n" .
			                 "    AddType text/xml .html\n".
			                 "</IfModule>";
        	}
        	
        	if( isset($var['headers']['Link']) && !empty($var['headers']['Link']) ) {
        		
        		$links = (is_array($var['headers']['Link'])) ? $var['headers']['Link'] : array($var['headers']['Link']);
        		
        		$htaccess .= "<IfModule mod_headers.c>\n";
        		foreach ($links as $link){
					$htaccess .= "    Header add Link ".$link."\n";
        		}
				$htaccess .= "</IfModule>";
        	}
        
        	if( !empty($htaccess) ){
        		file_put_contents(dirname($path) . '/.htaccess', $htaccess);
        	}
        }

        return true;
    }
```

**Step 2: parsing the wp output buffer and add push headers to the response** 

We can work in `/w3-total-cache-fixed/lib/W3/PgCache.php` on function `_can_cache_buffer(&$buffer, $has_dynamic)`, we can add the link headers, eg.:

```php
        if( $this->_config->get_boolean('pgcache.http2push.enabled', true) ){
            if( $this->_config->get_boolean('pgcache.http2push.styles', true) ){
                if( preg_match_all("/href=['\"](?P<css>([^'\"]+?\.css)[^'\"]*)/", $buffer, $matches) ){
                    $styles = $matches["css"];
                    foreach($styles as $style){
                        $headers['Link'][] = "<".$style.">;rel=preload;as=style";
                    }
                }
            }
            if( $this->_config->get_boolean('pgcache.http2push.scripts', true) ){
                if( preg_match_all("/src=['\"](?P<js>([^'\"]+?\.js)[^'\"]*)/", $buffer, $matches) ){
                    $scripts = $matches["js"];
                    foreach($scripts as $script){
                        $headers['Link'][] = "<".$script.">;rel=preload;as=script";
                    }
                }
            }
        }
```

and abjust _headers for expect array of headers:

```php
/**
     * Sends headers
     *
     * @param array $headers
     * @return boolean
     */
    function _headers($headers) {
        if (!headers_sent()) {
            foreach ($headers as $name => $value) {
            	if( is_array($value) ){
            		foreach ($value as $subvalue){
		                $header = ($name == 'Status' ? $subvalue : $name . ': ' . $subvalue);
		                @header($header);
            		}
            	} else {
            		$header = ($name == 'Status' ? $value : $name . ': ' . $value);
            		@header($header);
            	}
            }

            return true;
        }

        return false;
    }
```
this require a previous parsing of `$buffer` for find all css and/or js to push, parsing based on user choice by admin settings, eg.:

- allow/disallow push, remember push work only if webserver support http2 and website has ssl certificate.
- set which assets push: js, css, both (+ other media like image?)
- set max assets to push: to many assets has a performance degradation
- set max length for the headers: some cdn/web servers has a size limit
- set max size for the assets to push: http2 push has a performance degradation is the file to push is to big

**Comment by nigrosimone** [comment]
this is my idea https://github.com/szepeviktor/w3-total-cache-fixed/tree/http2-push it require a lot of more work for a seriously implementation.

**Comment by superpoincare** [comment]
Nice! 

Unfortunately don't have a server to test it ... The feature will be available soon says my host. 

Understand implementation will require a lot more effort. If the filename changes, the htaccess should also change and the usual stuff associated with html - such as garbage collection would need to run on this as well, for example. 

Just looking at your code ... you are creating htaccess for every html file created in the same cache folder: is that right? Which looks perfect since each each html page may have its own minified css.

**Comment by superpoincare** [comment]
Should also add that push is the future. Wordpress announced that starting Jan 1, 2017, it will only recommend sites which have https enabled by default. 

Push doesn't require https but all browsers have implemented it so that it will only work for https. Once servers enable https by default, they may also upgrade other things. I still have Apache 2.2 but might be upgraded to 2.4 which supports push.

**Comment by nigrosimone** [comment]
> you are creating htaccess for every html file created in the same cache folder: is that right? Which looks perfect since each each html page may have its own minified css.

Yes!

**Comment by nigrosimone** [comment]
closing for now... this features is too new for a stable implementation.

**Comment by nigrosimone** [comment]
@vijayaraghavanramanan sorry for the long delay (i haven't forgotten your idea ;) ). 

If you are yet interessed, the 0.9.5.x version also have a http 2 push feature but only for memory cache modules like memcache, redis, apcu.

 I have also implemented this features for page cache enhanced in the pr https://github.com/szepeviktor/w3-total-cache-fixed/pull/433.
